make `slugify()` produce valid url slugs

lukeaskew opened the pull request:
`some-headling` is a more user-friendly url than `SomeHeading`

This PR includes string.js's slugify function to handle all the complexities of slugifying strings.

lukeaskew committed:
make `slugify()` produce valid url slugs

adam-p commented:
I tried string.js when implementing this, and found it lacking: it's totally lacking extended character support (unless I'm missing something). So you get this:

``` js
S('こんにちは').slugify().s
> ""
```

...Which is unacceptable for my Markdown Here purposes.

---

The other (soft-ish) Markdown Here requirement that your PR maybe goes against is that users need to know what the actual anchor name will be, and it's best that they can do the conversion easily and by eye. 

(On the other hand, it's not a hard-and-fast requirement. And the alternate slugification could be optional.)

lukeaskew committed:
allow extended characters in slugification

lukeaskew commented:
Got it. Updated to use a less strict lib like https://github.com/blakeembrey/change-case to allow for extended characters.

Not too familiar with Markdown Here. Does this PR still go against how users predict urls?

adam-p commented:
That's certainly better, but... I still prefer the existing one. How about instead we add a slugifier (function) option? Then you can use whatever suits you.

Markdown Here: [project](https://github.com/adam-p/markdown-here), [site](http://markdown-here.com/). It's a browser extension for writing email in Markdown.

lukeaskew committed:
Add `slugify` option.

revert previous attempts at improving `slugifiy()` in favor of a user-defined function

lukeaskew commented:
Fair enough. More user control is better.

Added the function option, but made a small change the argument order - passing the string first so people can be a little more terse with function definitions.

adam-p commented:
Looks good. Merged. Thank you!
